File: submission/capture_the_flag.py
"""Capture the Flag - Mathathon submission. Paste this whole file into the editor.

Game: 29x29 grid, two teams of two. Grab the ENEMY flag and carry it into your
own territory to win. You get caught (die, respawn after 30 turns) if you stand
in enemy territory next to a living opponent. Hydration drains every turn
(2/turn at home, 1/turn elsewhere); the central 5x5 oasis refills it to 140 and
is also a safe haven (neutral - nobody can be caught there).

Strategy
  * One bot instance per player; the teammate runs this same code (a clone).
    Roles split with NO communication: a lexicographic tiebreak on positions,
    committed the moment the two clones diverge. One ATTACKS (goes for the
    enemy flag), one DEFENDS (guards the home front, intercepts intruders/carriers).
  * Movement = greedy descent of BFS distance fields. Static fields (enemy
    flag, our flag, our territory, oasis) are computed once at round 0; at
    most one extra BFS per round (chasing a moving enemy). Cheap and safe on
    any time budget.
  * Danger model: a cell in enemy territory within Chebyshev 1 of a living
    enemy is where you get caught - avoided. If progress stalls too long the
    attacker turns "desperate" and pushes through anyway (a draw beats a
    timid stalemate; a death just costs a respawn, not a forfeit).
  * Refuels at the oasis when hydration would not otherwise last.

Robustness: never writes stderr (the platform treats stderr as a forfeit);
every round prints exactly one of u/d/l/r/s and flushes; all decision logic is
wrapped so a bug emits the last move instead of crashing.
"""
import sys
import random

SIZE = 29
N = SIZE * SIZE
INF = 1 << 29
DIRS = (("u", 0, -1), ("d", 0, 1), ("l", -1, 0), ("r", 1, 0))


def in_oasis(x, y):
    return 12 <= x <= 16 and 12 <= y <= 16


def territory(x, y):
    """'B' = blue home, 'R' = red home, 'N' = neutral (oasis + row 14)."""
    if 12 <= x <= 16 and 12 <= y <= 16:
        return "N"
    if y <= 13:
        return "B"
    if y >= 15:
        return "R"
    return "N"


def bfs(sources, adj):
    """Multi-source BFS over free cells; returns a flat distance array."""
    dist = [INF] * N
    q = []
    for s in sources:
        if 0 <= s < N and dist[s] == INF:
            dist[s] = 0
            q.append(s)
    head = 0
    while head < len(q):
        c = q[head]
        head += 1
        nd = dist[c] + 1
        for nb in adj[c]:
            if dist[nb] == INF:
                dist[nb] = nd
                q.append(nb)
    return dist


class Bot:
    def __init__(self):
        self.ready = False
        self.free = None
        self.adj = None
        self.blue = True
        self.my_flag = 0
        self.enemy_flag = 0
        self.my_terr = "B"
        self.enemy_terr = "R"
        self.d_enemyflag = None
        self.d_myterr = None
        self.d_oasis = None
        self.d_myflag = None
        self.d_guard = None
        self.role = None            # committed 'A' / 'D'
        self.refilling = False
        self.stall = 0
        self.last_tgt_id = None
        self.last_tgt_val = INF
        self.last_move = "s"

    # ---- one-time setup -------------------------------------------------
    def setup_board(self, board_text):
        free = bytearray(N)
        L = len(board_text)
        for i in range(N):
            free[i] = 0 if (i < L and board_text[i] == "#") else 1
        adj = [()] * N
        for y in range(SIZE):
            for x in range(SIZE):
                i = y * SIZE + x
                if not free[i]:
                    continue
                nb = []
                for _, dx, dy in DIRS:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < SIZE and 0 <= ny < SIZE and free[ny * SIZE + nx]:
                        nb.append(ny * SIZE + nx)
                adj[i] = tuple(nb)
        self.free = free
        self.adj = adj

    def setup_team(self, mx, my):
        # Blue starts/respawns at y=0, red at y=28 -> y tells the team.
        self.blue = (my <= 14)
        if self.blue:
            self.my_flag = 0                       # (0, 0)
            self.enemy_flag = 28 * SIZE + 28       # (28, 28)
            self.my_terr, self.enemy_terr = "B", "R"
        else:
            self.my_flag = 28 * SIZE + 28
            self.enemy_flag = 0
            self.my_terr, self.enemy_terr = "R", "B"
        free, adj = self.free, self.adj
        self.d_enemyflag = bfs([self.enemy_flag], adj)
        self.d_myflag = bfs([self.my_flag], adj)
        self.d_oasis = bfs(
            [y * SIZE + x for y in range(12, 17) for x in range(12, 17)
             if free[y * SIZE + x]], adj)
        self.d_myterr = bfs(
            [y * SIZE + x for y in range(SIZE) for x in range(SIZE)
             if free[y * SIZE + x] and territory(x, y) == self.my_terr], adj)
        guard = []
        for y in range(SIZE):
            for x in range(SIZE):
                i = y * SIZE + x
                if not free[i]:
                    continue
                if self.my_terr == "B":
                    if y in (12, 13) and 2 <= x <= 26:
                        guard.append(i)
                elif y in (15, 16) and 2 <= x <= 26:
                    guard.append(i)
        self.d_guard = bfs(guard or [self.my_flag], adj)
        self.ready = True

    # ---- helpers --------------------------------------------------------
    def _danger(self, x, y, enemies):
        """True if cell (x,y) is where I could be caught next turn."""
        if territory(x, y) != self.enemy_terr:
            return False
        for ex, ey, _, _ in enemies:
            if max(abs(ex - x), abs(ey - y)) <= 1:
                return True
        return False

    @staticmethod
    def _min_cheb(x, y, enemies):
        m = 99
        for ex, ey, _, _ in enemies:
            d = max(abs(ex - x), abs(ey - y))
            if d < m:
                m = d
        return m

    # ---- per-round decision --------------------------------------------
    def decide(self, v):
        players = [(v[i], v[i + 1], v[i + 2], v[i + 3]) for i in range(0, 16, 4)]
        mx, my, mh, mf = players[0]
        if mx < 0:                                  # dead / respawning
            self.stall = 0
            self.refilling = False
            return "s"
        if not self.ready:
            self.setup_team(mx, my)

        mate = players[1]
        mate_alive = mate[0] >= 0
        enemies = [e for e in (players[2], players[3]) if e[0] >= 0]
        my_idx = my * SIZE + mx
        i_carry = (mf == 1)
        mate_carry = mate_alive and mate[3] == 1
        carrier = next((e for e in enemies if e[3] == 1), None)

        # ----- role assignment (no communication) -----
        chase_field = None
        provisional = False
        if i_carry:
            role = "A"
        elif carrier is not None:
            chase_field = bfs([carrier[1] * SIZE + carrier[0]], self.adj)
            my_d = chase_field[my_idx]
            mate_d = (chase_field[mate[1] * SIZE + mate[0]]
                      if mate_alive else INF)
            mine_closer = (my_d < mate_d or
                           (my_d == mate_d and (mx, my) < (mate[0], mate[1])))
            role = "D" if (not mate_alive or mine_closer) else "A"
        elif mate_carry:
            role = "D"
        elif not mate_alive:
            role = "A"
        elif self.role is not None:
            role = self.role                        # committed - sticky
        elif (mx, my) != (mate[0], mate[1]):
            role = "A" if (mx, my) < (mate[0], mate[1]) else "D"
            self.role = role
        else:
            role = "A"                              # both on same cell yet
            provisional = True

        # ----- pick the target distance field -----
        do = self.d_oasis[my_idx]
        if in_oasis(mx, my):
            self.refilling = False
        elif do != INF and mh < 2 * do + 14:
            self.refilling = True

        if self.refilling and not in_oasis(mx, my):
            tgt, tgt_id, goal = self.d_oasis, "oasis", (14, 14)
        elif role == "A":
            if i_carry:
                tgt, tgt_id = self.d_myterr, "home"
                goal = (self.my_flag % SIZE, self.my_flag // SIZE)
            else:
                tgt, tgt_id = self.d_enemyflag, "eflag"
                goal = (self.enemy_flag % SIZE, self.enemy_flag // SIZE)
        else:                                       # defender
            intruder = carrier
            if intruder is None:
                best = INF
                for e in enemies:
                    if territory(e[0], e[1]) == self.my_terr:
                        ds = self.d_myflag[e[1] * SIZE + e[0]]
                        if ds < best:
                            best, intruder = ds, e
            if intruder is not None:
                if chase_field is not None and intruder is carrier:
                    tgt = chase_field
                else:
                    tgt = bfs([intruder[1] * SIZE + intruder[0]], self.adj)
                tgt_id, goal = "chase", (intruder[0], intruder[1])
            else:
                tgt, tgt_id = self.d_guard, "guard"
                goal = (14, 13 if self.blue else 15)

        # ----- stall tracking -> desperate push -----
        cur = tgt[my_idx]
        if tgt_id != self.last_tgt_id or cur < self.last_tgt_val or cur == 0:
            self.stall = 0
        else:
            self.stall += 1
        self.last_tgt_id, self.last_tgt_val = tgt_id, cur
        desperate = self.stall > 25

        # ----- candidate moves -----
        cands = []
        for letter, dx, dy in DIRS:
            nx, ny = mx + dx, my + dy
            if 0 <= nx < SIZE and 0 <= ny < SIZE and self.free[ny * SIZE + nx]:
                cands.append((letter, nx, ny))
        cands.append(("s", mx, my))

        # Provisional clones share a cell - move randomly (but not backwards)
        # so the two diverge; positions then commit the lexicographic roles.
        if provisional:
            use_man = (cur == INF)
            gx, gy = goal
            forward = []
            for letter, nx, ny in cands:
                ni = ny * SIZE + nx
                dv = abs(nx - gx) + abs(ny - gy) if use_man else tgt[ni]
                if dv <= cur:
                    forward.append(letter)
            self.last_move = random.choice(forward or [c[0] for c in cands])
            return self.last_move

        # ----- score candidates -----
        use_man = (cur == INF)
        gx, gy = goal
        here_danger = self._danger(mx, my, enemies)
        best, best_key = "s", None
        for letter, nx, ny in cands:
            ni = ny * SIZE + nx
            if use_man:
                dval = abs(nx - gx) + abs(ny - gy)
            else:
                dv = tgt[ni]
                dval = dv if dv != INF else 9000 + abs(nx - gx) + abs(ny - gy)
            danger = self._danger(nx, ny, enemies) and not desperate
            terr_pen = (role == "D" and territory(nx, ny) == self.enemy_terr)
            safe = 0 if (danger or terr_pen) else 1
            if here_danger:                         # escape first
                key = (safe, self._min_cheb(nx, ny, enemies), -dval,
                       random.random())
            else:                                   # progress first
                key = (safe, -dval, random.random())
            if best_key is None or key > best_key:
                best_key, best = key, letter
        self.last_move = best
        return best


def main():
    bot = Bot()
    first = True
    for raw in sys.stdin:
        if first:
            first = False
            bot.setup_board(raw.strip())
            continue
        if not raw.strip():
            continue
        nums = []
        for tok in raw.split():
            try:
                nums.append(int(tok))
            except ValueError:
                pass
        if len(nums) < 16:
            sys.stdout.write((bot.last_move or "s") + "\n")
            sys.stdout.flush()
            continue
        try:
            move = bot.decide(nums[:16])
        except Exception:
            move = bot.last_move or "s"
        sys.stdout.write(move + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()

File: submission/capture_the_flag.cpp
// Capture the Flag - Mathathon C++ submission.
//
// 29x29 grid, two teams of two. Read the board on the first line, then each
// round read 16 ints: me, teammate, enemy0, enemy1 as x y hydration carrying.
// Print exactly one of u/d/l/r/s every round.

#include <algorithm>
#include <array>
#include <chrono>
#include <cstdint>
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>
#ifdef _WIN32
#include <process.h>
#else
#include <unistd.h>
#endif

using namespace std;

static constexpr int SIZE = 29;
static constexpr int N = SIZE * SIZE;
static constexpr int INF = 1 << 28;

struct Dir {
    char c;
    int dx;
    int dy;
};

static constexpr array<Dir, 4> DIRS{{
    {'u', 0, -1}, {'d', 0, 1}, {'l', -1, 0}, {'r', 1, 0}
}};

struct Player {
    int x = -1, y = -1, h = 0, f = 0;
};

static bool inOasis(int x, int y) {
    return 12 <= x && x <= 16 && 12 <= y && y <= 16;
}

static char territory(int x, int y) {
    if (inOasis(x, y)) return 'N';
    if (y <= 13) return 'B';
    if (y >= 15) return 'R';
    return 'N';
}

static int idx(int x, int y) {
    return y * SIZE + x;
}

struct Bot {
    bool boardReady = false;
    bool ready = false;
    array<unsigned char, N> freeCell{};
    array<vector<int>, N> adj{};

    bool blue = true;
    int myFlag = 0;
    int enemyFlag = 0;
    char myTerr = 'B';
    char enemyTerr = 'R';

    vector<int> dEnemyFlag, dMyTerr, dOasis, dMyFlag, dGuard;

    char committedRole = 0;
    bool refilling = false;
    int stall = 0;
    string lastTargetId;
    int lastTargetVal = INF;
    char lastMove = 's';
    uint32_t rng = 2463534242u;

    Bot() {
        auto now = chrono::high_resolution_clock::now().time_since_epoch().count();
        uintptr_t addr = reinterpret_cast<uintptr_t>(this);
        rng ^= (uint32_t)now;
        rng ^= (uint32_t)(now >> 32);
        rng ^= (uint32_t)addr;
#ifdef _WIN32
        rng ^= (uint32_t)_getpid() * 2654435761u;
#else
        rng ^= (uint32_t)getpid() * 2654435761u;
#endif
    }

    uint32_t rnd() {
        rng ^= rng << 13;
        rng ^= rng >> 17;
        rng ^= rng << 5;
        return rng;
    }

    void setupBoard(const string& board) {
        for (int i = 0; i < N; ++i) {
            freeCell[i] = (i < (int)board.size() && board[i] == '#') ? 0 : 1;
        }
        for (int y = 0; y < SIZE; ++y) {
            for (int x = 0; x < SIZE; ++x) {
                int here = idx(x, y);
                adj[here].clear();
                if (!freeCell[here]) continue;
                for (const Dir& d : DIRS) {
                    int nx = x + d.dx, ny = y + d.dy;
                    if (0 <= nx && nx < SIZE && 0 <= ny && ny < SIZE &&
                        freeCell[idx(nx, ny)]) {
                        adj[here].push_back(idx(nx, ny));
                    }
                }
            }
        }
        boardReady = true;
    }

    vector<int> bfs(const vector<int>& sources) const {
        vector<int> dist(N, INF);
        queue<int> q;
        for (int s : sources) {
            if (0 <= s && s < N && freeCell[s] && dist[s] == INF) {
                dist[s] = 0;
                q.push(s);
            }
        }
        while (!q.empty()) {
            int c = q.front();
            q.pop();
            int nd = dist[c] + 1;
            for (int nb : adj[c]) {
                if (dist[nb] == INF) {
                    dist[nb] = nd;
                    q.push(nb);
                }
            }
        }
        return dist;
    }

    void setupTeam(int my) {
        blue = (my <= 14);
        if (blue) {
            myFlag = idx(0, 0);
            enemyFlag = idx(28, 28);
            myTerr = 'B';
            enemyTerr = 'R';
        } else {
            myFlag = idx(28, 28);
            enemyFlag = idx(0, 0);
            myTerr = 'R';
            enemyTerr = 'B';
        }

        dEnemyFlag = bfs({enemyFlag});
        dMyFlag = bfs({myFlag});

        vector<int> oasis, myTerrCells, guard;
        for (int y = 0; y < SIZE; ++y) {
            for (int x = 0; x < SIZE; ++x) {
                int p = idx(x, y);
                if (!freeCell[p]) continue;
                if (inOasis(x, y)) oasis.push_back(p);
                if (territory(x, y) == myTerr) myTerrCells.push_back(p);

                // Idle defender waits near our front line, still in home
                // territory, so intruders are caught soon after crossing.
                if (myTerr == 'B') {
                    if ((y == 12 || y == 13) && 2 <= x && x <= 26) guard.push_back(p);
                } else {
                    if ((y == 15 || y == 16) && 2 <= x && x <= 26) guard.push_back(p);
                }
            }
        }
        if (guard.empty()) guard.push_back(myFlag);
        dOasis = bfs(oasis);
        dMyTerr = bfs(myTerrCells);
        dGuard = bfs(guard);
        ready = true;
    }

    bool danger(int x, int y, const vector<Player>& enemies) const {
        if (territory(x, y) != enemyTerr) return false;
        for (const Player& e : enemies) {
            if (max(abs(e.x - x), abs(e.y - y)) <= 1) return true;
        }
        return false;
    }

    static int minCheb(int x, int y, const vector<Player>& enemies) {
        int best = 99;
        for (const Player& e : enemies) {
            best = min(best, max(abs(e.x - x), abs(e.y - y)));
        }
        return best;
    }

    char decide(const array<int, 16>& v) {
        array<Player, 4> p;
        for (int i = 0; i < 4; ++i) {
            p[i] = {v[i * 4], v[i * 4 + 1], v[i * 4 + 2], v[i * 4 + 3]};
        }
        Player me = p[0];
        if (me.x < 0) {
            stall = 0;
            refilling = false;
            return 's';
        }
        if (!ready) setupTeam(me.y);

        Player mate = p[1];
        bool mateAlive = mate.x >= 0;
        vector<Player> enemies;
        if (p[2].x >= 0) enemies.push_back(p[2]);
        if (p[3].x >= 0) enemies.push_back(p[3]);

        int myIdx = idx(me.x, me.y);
        bool iCarry = me.f == 1;
        bool mateCarry = mateAlive && mate.f == 1;
        int carrierIndex = -1;
        for (int i = 0; i < (int)enemies.size(); ++i) {
            if (enemies[i].f == 1) {
                carrierIndex = i;
                break;
            }
        }

        vector<int> chaseField;
        bool hasChaseField = false;
        bool provisional = false;
        char role;
        if (iCarry) {
            role = 'A';
        } else if (carrierIndex >= 0) {
            const Player& c = enemies[carrierIndex];
            chaseField = bfs({idx(c.x, c.y)});
            hasChaseField = true;
            int myD = chaseField[myIdx];
            int mateD = mateAlive ? chaseField[idx(mate.x, mate.y)] : INF;
            bool mineCloser = myD < mateD ||
                (myD == mateD && pair<int, int>(me.x, me.y) < pair<int, int>(mate.x, mate.y));
            role = (!mateAlive || mineCloser) ? 'D' : 'A';
        } else if (mateCarry) {
            role = 'D';
        } else if (!mateAlive) {
            role = 'A';
        } else if (committedRole) {
            role = committedRole;
        } else if (me.x != mate.x || me.y != mate.y) {
            role = pair<int, int>(me.x, me.y) < pair<int, int>(mate.x, mate.y) ? 'A' : 'D';
            committedRole = role;
        } else {
            role = 'A';
            provisional = true;
        }

        int doasis = dOasis[myIdx];
        if (inOasis(me.x, me.y)) {
            refilling = false;
        } else if (doasis != INF && me.h < 2 * doasis + 14) {
            refilling = true;
        }

        const vector<int>* target = nullptr;
        string targetId;
        int goalX = 14, goalY = 14;

        if (refilling && !inOasis(me.x, me.y)) {
            target = &dOasis;
            targetId = "oasis";
            goalX = goalY = 14;
        } else if (role == 'A') {
            if (iCarry) {
                target = &dMyTerr;
                targetId = "home";
                goalX = myFlag % SIZE;
                goalY = myFlag / SIZE;
            } else {
                target = &dEnemyFlag;
                targetId = "eflag";
                goalX = enemyFlag % SIZE;
                goalY = enemyFlag / SIZE;
            }
        } else {
            int intruderIndex = carrierIndex;
            if (intruderIndex < 0) {
                int best = INF;
                for (int i = 0; i < (int)enemies.size(); ++i) {
                    const Player& e = enemies[i];
                    if (territory(e.x, e.y) == myTerr) {
                        int ds = dMyFlag[idx(e.x, e.y)];
                        if (ds < best) {
                            best = ds;
                            intruderIndex = i;
                        }
                    }
                }
            }
            if (intruderIndex >= 0) {
                const Player& intr = enemies[intruderIndex];
                if (hasChaseField && intruderIndex == carrierIndex) {
                    target = &chaseField;
                } else {
                    chaseField = bfs({idx(intr.x, intr.y)});
                    hasChaseField = true;
                    target = &chaseField;
                }
                targetId = "chase";
                goalX = intr.x;
                goalY = intr.y;
            } else {
                target = &dGuard;
                targetId = "guard";
                goalX = blue ? 14 : 14;
                goalY = blue ? 13 : 15;
            }
        }

        int cur = (*target)[myIdx];
        if (targetId != lastTargetId || cur < lastTargetVal || cur == 0) {
            stall = 0;
        } else {
            ++stall;
        }
        lastTargetId = targetId;
        lastTargetVal = cur;
        bool desperate = stall > 25;

        struct Cand { char c; int x; int y; };
        vector<Cand> cands;
        for (const Dir& d : DIRS) {
            int nx = me.x + d.dx, ny = me.y + d.dy;
            if (0 <= nx && nx < SIZE && 0 <= ny && ny < SIZE && freeCell[idx(nx, ny)]) {
                cands.push_back({d.c, nx, ny});
            }
        }
        cands.push_back({'s', me.x, me.y});

        bool useMan = cur == INF;
        if (provisional) {
            vector<char> forward;
            vector<char> legal;
            for (const Cand& c : cands) {
                legal.push_back(c.c);
                int dval = useMan ? abs(c.x - goalX) + abs(c.y - goalY) : (*target)[idx(c.x, c.y)];
                if (dval <= cur) forward.push_back(c.c);
            }
            const vector<char>& opts = forward.empty() ? legal : forward;
            lastMove = opts[rnd() % opts.size()];
            return lastMove;
        }

        bool hereDanger = danger(me.x, me.y, enemies);
        char bestMove = 's';
        array<int, 4> bestKey{{-INF, -INF, -INF, -INF}};
        for (const Cand& c : cands) {
            int ni = idx(c.x, c.y);
            int dval;
            if (useMan) {
                dval = abs(c.x - goalX) + abs(c.y - goalY);
            } else {
                int dv = (*target)[ni];
                dval = (dv != INF) ? dv : 9000 + abs(c.x - goalX) + abs(c.y - goalY);
            }
            bool badDanger = danger(c.x, c.y, enemies) && !desperate;
            bool terrPenalty = role == 'D' && territory(c.x, c.y) == enemyTerr;
            int safe = (badDanger || terrPenalty) ? 0 : 1;
            int jitter = (int)(rnd() & 1023);
            array<int, 4> key;
            if (hereDanger) {
                key = {{safe, minCheb(c.x, c.y, enemies), -dval, jitter}};
            } else {
                key = {{safe, -dval, jitter, 0}};
            }
            if (key > bestKey) {
                bestKey = key;
                bestMove = c.c;
            }
        }
        lastMove = bestMove;
        return bestMove;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    Bot bot;
    string line;
    if (!getline(cin, line)) return 0;
    bot.setupBoard(line);

    while (getline(cin, line)) {
        if (line.empty()) continue;
        array<int, 16> vals{};
        istringstream iss(line);
        int n = 0;
        while (n < 16 && (iss >> vals[n])) ++n;
        char move = bot.lastMove;
        if (n >= 16) {
            try {
                move = bot.decide(vals);
            } catch (...) {
                move = bot.lastMove;
            }
        }
        cout << move << '\n' << flush;
    }
    return 0;
}
